src/protocol/events.rs

```diff
@@ -52,481 +49,321 @@ gen_rep!(
+            _ => unreachable!(),
         }
     }
 
     #[getter]
-    fn src_version(_self: PyRef<'_, Self>) -> PyResult<u64> {
-        match &_self.as_ref().0 {
-            events_listen::Rep::Ok(APIEvent::RealmVlobsUpdated { src_version, .. }) => {
-                Ok(*src_version)
-            }
-            _ => Err(PyAttributeError::new_err("")),
+    fn src_version(_self: PyRef<'_, Self>) -> u64 {
+        match &_self.into_super().as_ref().0 {
+            events_listen::Rep::Ok(APIEvent::RealmVlobsUpdated { src_version, .. }) => *src_version,
+            _ => unreachable!(),
         }
     }
 }
 
-#[pyclass(extends = EventsListenRep)]
+#[pyclass(extends = EventsListenRepOk)]
 #[derive(Clone)]
 pub(crate) struct EventsListenRepOkRealmRolesUpdated;
 
 #[pymethods]
 impl EventsListenRepOkRealmRolesUpdated {
     #[new]
-    fn new(realm_id: RealmID, role: &PyAny) -> PyResult<(Self, EventsListenRep)> {
-        let r = py_to_rs_realm_role(role)?;
-        Ok((
-            Self,
-            EventsListenRep(events_listen::Rep::Ok(APIEvent::RealmRolesUpdated {
-                realm_id: realm_id.0,
-                role: r,
-            })),
-        ))
+    fn new(realm_id: RealmID, role: Option<RealmRole>) -> PyClassInitializer<Self> {
+        let event = events_listen::APIEvent::RealmRolesUpdated {
+            realm_id: realm_id.0,
+            role: role.map(|x| x.0),
+        };
+        PyClassInitializer::from(EventsListenRepOk::new(event)).add_subclass(Self)
     }
 
     #[getter]
-    fn realm_id(_self: PyRef<'_, Self>) -> PyResult<RealmID> {
-        match &_self.as_ref().0 {
+    fn realm_id(_self: PyRef<'_, Self>) -> RealmID {
+        match &_self.into_super().as_ref().0 {
             events_listen::Rep::Ok(APIEvent::RealmRolesUpdated { realm_id, .. }) => {
-                Ok(RealmID(*realm_id))
+                RealmID(*realm_id)
             }
-            _ => Err(PyAttributeError::new_err("")),
+            _ => unreachable!(),
         }
     }
 
     #[getter]
-    fn role(_self: PyRef<'_, Self>) -> PyResult<Option<&'static str>> {
-        match &_self.as_ref().0 {
-            events_listen::Rep::Ok(APIEvent::RealmRolesUpdated { role, .. }) => match role {
-                Some(libparsec::types::RealmRole::Owner) => Ok(Some("OWNER")),
-                Some(libparsec::types::RealmRole::Manager) => Ok(Some("MANAGER")),
-                Some(libparsec::types::RealmRole::Contributor) => Ok(Some("CONTRIBUTOR")),
-                Some(libparsec::types::RealmRole::Reader) => Ok(Some("READER")),
-                None => Ok(None),
-            },
-            _ => Err(PyAttributeError::new_err("")),
+    fn role(_self: PyRef<'_, Self>) -> Option<RealmRole> {
+        match &_self.into_super().as_ref().0 {
+            events_listen::Rep::Ok(APIEvent::RealmRolesUpdated { role, .. }) => {
+                role.and_then(|x| Some(RealmRole(x)))
```

Comment:
```suggestion
                role.map(RealmRole)
```

Reply:
strange than clippy didn't catch that :/